File: 6_collision/collision_sol.py
#!/usr/bin/env python
# Copyright 2017 J Carruthers jbc@bu.edu
#
"""
assignment 6 solution
"""

import sys
import numpy as np

RADIUS = 5


class Circle():
    "model a circle with linear velocity"

    def __init__(self, name, position, velocity):
        self.name = name
        self.pos = position
        self.vel = velocity

    def __str__(self):
        return "{} {} {} {} {}".format(self.name,
                                       round(self.pos[0], 10),
                                       round(self.pos[1], 10),
                                       round(self.vel[0], 10),
                                       round(self.vel[1], 10))

    def move(self, deltat):
        "move the circle"
        self.pos = self.pos + deltat * self.vel

    def collide_time(self, two):
        """predict the time of collision between self and two,
           Inf if no valid collision"""
        delr = self.pos - two.pos
        delv = self.vel - two.vel

        a_co = delv @ delv
        if a_co == 0:
            return np.Inf

        dr2 = delr @ delr
        b_co = 2 * delr @ delv
        rootarg = b_co * b_co - 4 * a_co * (dr2 - 4 * RADIUS * RADIUS)
        if rootarg < 0:
            return np.Inf
        root = rootarg**0.5

        colltime = (-b_co - root) / (2 * a_co)
        if colltime == 0 and (delr @ delv) < 0:
            return 0
        elif colltime > 0:
            return colltime

        return np.Inf


def get_report_times():
    "process the report times from command line"
    all_times = []
    try:
        for timest in sys.argv[1:]:
            time = float(timest)
            if time >= 0:
                all_times.append(time)
    except ValueError:
        return None
    return sorted(all_times)


def get_circle_data():
    "read in the circle data"
    circles = []
    for line in sys.stdin:
        try:
            name, posx, posy, velx, vely = line.split()
            pos = np.array((float(posx), float(posy)))
            vel = np.array((float(velx), float(vely)))
            circles.append(Circle(name, pos, vel))
        except ValueError:
            return None

    return circles


def get_next_collision(cur_time, circles):
    "determine the next collision time for all circles"
    next_collision = np.Inf
    colliders = None, None
    for i, one in enumerate(circles):
        for j, two in enumerate(circles):
            if i >= j:
                continue
            colltime = one.collide_time(two)
            if colltime < next_collision:
                next_collision = colltime
                colliders = i, j

    return colliders, cur_time + next_collision


def update_position(circles, deltat):
    "move circles"
    for circ in circles:
        circ.move(deltat)


def print_report(cur_time, circles):
    "print time and position of all circles"
    print(cur_time)
    for circ in circles:
        print(circ)


def elastic_collision(one, two):
    """one and two are colliding. update their velocities"""
    hitvec = one.pos - two.pos
    alpha = (one.vel - two.vel) @ hitvec / (hitvec @ hitvec)

    assert alpha <= 0, "{} {}".format(one, two)
    two.vel = two.vel + alpha * hitvec
    one.vel = one.vel - alpha * hitvec


def main():
    "get simulation and report on elastic collisions"
    times = get_report_times()
    if not times:
        return 2

    circles = get_circle_data()
    if not circles:
        return 1

    cur_time = 0

    while True:
        # find the next collision time and circles involved.
        (one, two), c_time = get_next_collision(cur_time, circles)

        # print out all reports prior to collision
        while times and (times[0] <= c_time):
            update_position(circles, times[0] - cur_time)
            cur_time = times[0]
            print_report(cur_time, circles)
            del times[0]

        # stop when no more reports are due
        if not times:
            break

        # move all objects to the collision time
        update_position(circles, c_time - cur_time)
        cur_time = c_time

        # record the collision
        elastic_collision(circles[one], circles[two])

    return 0


if __name__ == '__main__':
    exit(main())

File: 6_collision/collision_sol.cpp
// Copyright 2017 Jeff Carruthers jbc@bu.edu
//
// Solution to collision

#include <vector>
#include <iostream>
#include <sstream>
#include <cmath>
#include <algorithm>
#include <string>

using std::vector;
using std::cout;
using std::cerr;
using std::stod;



typedef double numbers;

const numbers NO_COLLIDE_SENTINEL = -1;
const numbers RADIUS = 5;

struct Collision {
  numbers event_time;
  int one, two;
};

struct Stone {
  std::string name;
  numbers rx, ry, vx, vy;  // position and speed
};


void collide_two_stones(Stone *one, Stone *two) {
  numbers xx = two->rx - one->rx;
  numbers xy = two->ry - one->ry;
  numbers vx = two->vx - one->vx;
  numbers vy = two->vy - one->vy;

  numbers D2 = xx * xx + xy * xy;

  numbers dot1 = vx * xx + vy * xy;

  one->vy  +=  (dot1) * (xy) / D2;
  two->vy  +=  -(dot1) * (xy) / D2;

  one->vx  += (dot1) * (xx) / D2;
  two->vx  += -(dot1) * (xx) / D2;
}


void update_position(Stone *one, const numbers dt) {
  one->rx += one->vx * dt;
  one->ry += one->vy * dt;
}


numbers collide_time(const Stone* one, const Stone* two) {
  numbers a = two->rx - one->rx;
  numbers b = two->ry - one->ry;
  numbers c = two->vx - one->vx;
  numbers d = two->vy - one->vy;
  numbers A = c * c + d * d;
  numbers B = 2 * (a * c + b * d);

  numbers C = (a * a + b * b - 4 * RADIUS * RADIUS);

  numbers rootarg = B * B - 4 * A * C;

  if (rootarg < 0) return NO_COLLIDE_SENTINEL;


  numbers root = sqrt(rootarg);

  numbers t1 = (-B - root) / (2 * A);

  if ((t1 == 0) and (a * c + b * d < 0))
    return 0;
  else if (t1 > 0)
    return t1;

  return NO_COLLIDE_SENTINEL;
}


void stone_report(const vector<Stone> & stones, numbers curr_time) {
  cout << curr_time << '\n';

  for (int i = 0; i < stones.size() ; i++) {
    cout << stones[i].name << ' '
         << stones[i].rx << ' '
         << stones[i].ry << ' '
         << stones[i].vx << ' '
         << stones[i].vy << '\n';
  }
}

Collision get_next_collision(numbers curr_time, const vector<Stone> &stones) {
  numbers next_collision = 1e38;
  numbers ct, one, two;
  Collision c;

  for (int i = 0; i < stones.size() ; i++) {
    for (int j = i + 1; j < stones.size() ; j++) {
      ct = collide_time(&stones[i], &stones[j]);
      if ((ct != NO_COLLIDE_SENTINEL) and (ct < next_collision)) {
        next_collision = ct;
        c.one = i;
        c.two = j;
      }
    }
  }

  c.event_time = next_collision + curr_time;
  return c;
}


int main(int argc, char const *argv[]) {
  cout.precision(8);

  vector<numbers> report_times;
  vector<Stone> stones;
  Stone s;
  numbers curr_time, dt;
  int one, two;  // stones involved in the event.
  try {
    for (int i = 1; i < argc; i++) {
      numbers value = std::stod(argv[i]);
      if  (value >= 0)
        report_times.push_back(value);
    }
  } catch (...) {
    return 2;
  }

  if (report_times.empty())
    return 2;

  std::sort(begin(report_times), end(report_times));

  curr_time = 0;


  std::stringstream ss;

  std::string line, rest, val;

  while (std::getline(std::cin, line)) {
    ss << line;
    try {
      ss >> s.name;
      ss >> val;
      s.rx = std::stod(val);
      ss >> val;
      s.ry = std::stod(val);
      ss >> val;
      s.vx = std::stod(val);
      ss >> val;
      s.vy = std::stod(val);
      if (ss.fail()) throw 1;
      ss >> rest;
      if (not rest.empty()) throw 1;
    } catch (...) {
      return 1;
    }

    ss.str("");
    ss.clear();

    stones.push_back(s);
  }

  int  next_report_index = 0;

  Collision collision;

  while (true) {
    // find the next collision event and the colliding object

    collision = get_next_collision(curr_time, stones);

    // print out all report times prior to collision event.

    while ( (next_report_index < report_times.size()) and
            ( (collision.event_time != NO_COLLIDE_SENTINEL) and
              (collision.event_time >= report_times[next_report_index]))
          ) {
      dt = report_times[next_report_index] - curr_time;

      for (int i = 0; i < stones.size(); i++)
        update_position(&stones[i], dt);

      curr_time = report_times[next_report_index];

      stone_report(stones, curr_time);
      next_report_index++;
    }

    // Are we done?
    if (next_report_index == report_times.size())
      break;

    // move time to collision event
    dt  = collision.event_time - curr_time;
    curr_time = collision.event_time;

    for (int i = 0; i < stones.size(); i++)
      update_position(&stones[i], dt);


    collide_two_stones(&stones[collision.one], &stones[collision.two]);
  }
  return 0;
}
